extra_geom/base.py

```diff
@@ -791,3 +798,135 @@ def offset(self, shift, *, modules=np.s_[:], tiles=np.s_[:]):
+                             f'Last dimension was {rot.shape[-1]}')
+        if center is not None and np.asarray(center).shape != rot.shape:
+            raise ValueError('Rotation center must have same shape as angles')
+
+        ntiles = len(max((m for m in self.modules), key=len))
+        all_rot = np.zeros((len(self.modules), ntiles, 3), dtype=rot.dtype)
+        sel_rot = all_rot[modules, tiles]
+
+        refs = np.zeros((len(self.modules), ntiles, 3), dtype=np.float64)
+        sel_refs = refs[modules, tiles]
+
+        if center is None:
+            if rot.shape[:-1] == sel_rot.shape[:2] or tiles != np.s_[:]:
+                # Per-tile rotation
+                refs[modules, tiles] = np.array(
+                    [[t.centre() for t in m[tiles]]
+                    for m in self.modules[modules]]
+                )
+            elif rot.shape[:-1] == sel_rot.shape[:1] or modules != np.s_[:]:
+                # Per-module rotation
+                mods = []
+                for module in self.modules[modules]:
+                    center = np.mean([t.centre() for t in module], axis=0)
+                    mods.append(np.array([center for _ in module[tiles]]))
+                refs[modules, tiles] = np.array(mods)
+        else:
+            center = np.asarray(center)
+            if center.shape[:-1] == sel_refs.shape[:2]:
+                # Per-tile offsets
+                sel_refs[:] = center
+            elif center.shape[:-1] == sel_refs.shape[:1]:
+                # Per-module offsets - broadcast across tiles
+                sel_refs[:] = center[:, np.newaxis]
+            elif center.shape[:-1] == ():
+                # Single shift - broadcast across modules and tiles
+                sel_refs[:] = center
+            else:
+                raise ValueError(
+                    f"Got {center.shape[:-1]} coordinates. Expected either a "
+                    f"single coordinate (), a coordinate per module "
+                    f"{sel_refs.shape[:1]} or a coordinate per tile "
+                    f"{sel_refs.shape[:2]}"
+                )
+
+        if rot.shape[:-1] == sel_rot.shape[:2]:
+            # per-tile rotations
+            sel_rot[:] = rot
+        elif rot.shape[:-1] == sel_rot.shape[:1]:
+            # per-module rotations - broadcast across tiles
+            sel_rot[:] = rot[:, np.newaxis]
+        elif rot.shape[:-1] == ():
+            # single rotation - broadcast across modules and tiles
+            sel_rot[:] = rot
+        else:
+            raise ValueError(
+                f"Got {rot.shape[:-1]} coordinates. Expected either a single "
+                f"coordinate (), a coordinate per module {sel_rot.shape[:1]} "
+                f"or a coordinate per tile {sel_rot.shape[:2]}"
+            )
+
+        @lru_cache()
+        def _rot(rotations):
+            # generate rotation matrix for the given angles
+            if degrees:
+                w, p, k = np.deg2rad(rotations)
```

Comment:
I might do this conversion outside the inner function - it should be easy to apply it to the whole `rot` array, I think.